### ISSUE TITLE
[pdf_viewer] [pdf] `document.pages[controller.pageNumber]` throws Null Pointer Exception

### ISSUE THREAD

**Issue by ABausG** [syncfusion/flutter-widgets#1281]
I am trying to build a workaround for #1229 to manipulate a pdf.

Testing on a Real Android Phone (Pixel 3a)

Here are some snippets of what I'm doing right now. See code comments and explanation with issues I am having. It would be greatly appreciated if you could give some hints as to how I could adapt this as this really is a crucial feature as we want to enable selecting text and setting this as highlights.

```dart
final _controller = PdfViewerController();

Uint8List _bytes = getPdfBytes();

late PdfDocument _document;

Widget build(BuildContext context) {
  return SfPdfViewer.memory(
    _bytes,
    controller: _controller,
    onDocumentLoaded: (details) {
      _document = details.document;
    },
  );
}

Future<void> manipulate(Rect bounds, PdfColor color) async {
  final page = _document.pages[_controller.pageNumber];

  // Annotations are not added to the document
  final annotation = PdfRectangleAnnotation(
      bounds, 
      'Test Annotation',
      innerColor: color,
  );
  final annotations = page.annotations;
  annotations.add(annotation);
  
  page?.graphics.drawRectangle(bounds: bounds, brush: PdfSolidBrush(color));

  // Just Saving the document does not update the pdf. This always throws an OutOfMemory Error for me as the pdfs I'm dealing with are >100MBs in size
  await _document.save();
  
  // For some reason this does sometimes work and does not get an OutOfMemory
  setState(() {
    _bytes = await _document.save();
  }
}
```

Now to some more issues. When I run the `manipulate` function and have the Rectangle drawn. As expected and described in #1229 I get the grey visual and the loading bar at the top. Once the page is loaded I can see the rectangle drawn as expected. However if I then proceed to call the `manipulate` function again I get the following error when accessing `_document.pages[_controller.pageNumber]`. This means I can't update the pdf to add a second annotation.
```
E/flutter ( 8491): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Null check operator used on a null value
E/flutter ( 8491): #0      PdfDictionary.returnValue (package:syncfusion_flutter_pdf/src/pdf/implementation/primitives/pdf_dictionary.dart:125:14)
E/flutter ( 8491): #1      PdfDictionary.[] (package:syncfusion_flutter_pdf/src/pdf/implementation/primitives/pdf_dictionary.dart:60:46)
E/flutter ( 8491): #2      PdfPageCollection.count (package:syncfusion_flutter_pdf/src/pdf/implementation/pages/pdf_page_collection.dart:69:19)
E/flutter ( 8491): #3      PdfPageCollection._getParent (package:syncfusion_flutter_pdf/src/pdf/implementation/pages/pdf_page_collection.dart:409:17)
E/flutter ( 8491): #4      PdfPageCollection._returnValue (package:syncfusion_flutter_pdf/src/pdf/implementation/pages/pdf_page_collection.dart:238:43)
E/flutter ( 8491): #5      PdfPageCollection.[] (package:syncfusion_flutter_pdf/src/pdf/implementation/pages/pdf_page_collection.dart:62:37)
```

Crucially the `onDocumentLoaded` callback is not called again after updating the `bytes`

Another Issue is that the annotation is not shown. And even the described behaviour in https://help.syncfusion.com/flutter/pdf/working-with-annotations#rectangle-annotation 
> When you open it, it displays a pop-up window containing the text of the associated note.

Does not really allow me to fully customise that behaviour. Ideally there would be a callback like: `final void Function(Annotation) onAnnotationSelected` in the `SfPdfViewer` to full customise this behaviour. Alternatively a similar callback with a Graphic would be nice. In that case I think some sort of id should be added to the graphic. Similar to Markers on a Google Map

---

So to summarise I have the following question wishes. It would be great to get an answer/directions on if this is a possible way to workaround the missing annotations.

1. Display added annotations
2. Allow to manipulate a document multiple times: Get access to pages after updating
3. Update a `PdfDocument` inside `SfPdfViewer` without having save the document/to reload the whole bytes array. Especially for large documents this is a real issue due to memory constraints and load times
4. Detect Clicks of Annotations/Graphics to allow customising the behaviour.

**Comment by deepika134** [comment]
**Regarding displaying the added annotation and manipulating the document multiple times:**

We attempted to replicate the issue in the workaround sample, but the annotation is added properly, and the document load callback is triggered correctly when we reload the document. For your reference, we have attached the sample along with the output screen recording, which can be downloaded from the attachment. If you are still facing the same issue in the sample, kindly modify the issue to reproduce it in the sample and share a video or steps to replicate the issue. This will be helpful for us to provide a solution as quickly as possible.

**Regarding restricting the reloading of the document:**

Currently, we do not have support for text markup annotation in SfPdfViewer. In the workaround, it is not possible to add an annotation without reloading the document in SfPdfViewer. We have logged this as a feature and please track the status of the feature in the link below.

https://www.syncfusion.com/feedback/26331/support-for-text-markup-annotation

[Text_Markup.zip](https://github.com/syncfusion/flutter-widgets/files/11932947/Text_Markup.zip)

**Comment by ramkumarsync** [comment]
The support to add, remove, modify, and save text markup annotations in PDF files is available from the 2023 Volume 4 release. Please find the package link below. The available text markups are highlight, underline, strikethrough, and squiggly.

https://pub.dev/packages/syncfusion_flutter_pdfviewer/versions/24.1.41
